ls1intum/Artemis
# `Programming exercises`: Make export more robust

### Checklist
#### General


- [x] I tested **all** changes and their related features with **all** corresponding user types locally.
#### Server
#### Client


#### Changes affecting Programming Exercises
- [x] I tested **all** changes and their related features with **all** corresponding user types locally with the Atlassian Suite
- [x] I tested **all** changes and their related features with **all** corresponding user types locally with Jenkins and Gitlab.

### Motivation and Context
Closes #6935 
Up until now, the export failed if the problem statement was null. Unfortunately, it even failed silently.
Instead one should be able to export the programming exercise even if the problem statement is null.
### Description

- only include problem statement if is not null
- export no longer fails if the embedded files cannot be copied
-  client now shows an alert if the export fails.
-  We cannot show a specific exception message from the server because the response type is a blob, see https://github.com/angular/angular/issues/19148

### Steps for Testing

Best tested locally, as you ideally need access to the database.
Prerequisites:
- 1 Instructor
- 1 Programming exercise

1. Set the problem statement to null in the database
2. Try to export the exercise which should be successful
3. Check that the export contains everything as for another programming exercise for which the problem statement is not null except the problem statement markdown file.
4. include some files in the problem statement and save the PE
5. delete the files on the file system.
6. Export the PE 
7. Export should contain everything but the embedded files.
8. Try to generate a PE for which the export could fail and see if you get an alert on the client if it fails.



### Review Progress






#### Performance Review
#### Code Review
#### Manual Tests

### Test Coverage






| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [programming-exercise-instructor-exercise-download.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5229/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/programming/shared/actions/programming-exercise-instructor-exercise-download.component.ts.html) | 100% | ✅|
#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [ProgrammingExerciseExportService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5229/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.programming/ProgrammingExerciseExportService.html) | 85% | ✅ |
| [ProgrammingExerciseExportImportResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5229/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest/ProgrammingExerciseExportImportResource.html) | 95% | ✅|

### Screenshots


![image](https://github.com/ls1intum/Artemis/assets/84102468/496aa4d1-9945-4398-842f-369ffce26413)
